src/osp/Universe.h

```diff
@@ -224,112 +222,62 @@ struct UCompTransformTraj
     std::string m_name;
 };
 
-//struct ACompActivated
-//{
-//    // Index to [probably an active area] when activated
-//    Satellite m_area;
-//}
-
 struct UCompVelocity
 {
     Vector3 m_velocity;
 };
 
 struct UCompType
 {
-    // maybe use an iterator
-    ITypeSatellite* m_type{nullptr};
-};
-
-struct UCompActivatable
-{
-    // put something here some day
-    int m_dummy;
-};
-
-
-/**
- * A specific type of Satellite. A planet, star, vehicle, etc...
- */
-class ITypeSatellite
-{
-public:
-    virtual ~ITypeSatellite() = default;
-    virtual std::string get_name() = 0;
-
-    virtual void add(Satellite sat) = 0;
-    virtual void remove(Satellite sat) = 0;
+    TypeSatIndex m_type{TypeSatIndex::Invalid};
 };
 
-
 /**
- * CRTP to implement common functions of TypeSatellites. add and remove
- * functions will add or remove the component UCOMP_T
+ * Calls rUni.get_reg().emplace<UCOMP_T>(sat), and returns a tuple containing a
+ * reference to the UCOMP_T, or a blank tuple if UCOMP_T is a tag struct.
+ * @tparam UCOMP_T
+ * @param rUni [out] Universe to add new components to
+ * @param sat  [in]  Satellite that is part of rUni
  */
-template<typename TYPESAT_T, typename ... UCOMP_T>
-class CommonTypeSat : public ITypeSatellite
-{
-
-public:
-
-
-    CommonTypeSat(Universe& universe) : m_universe(universe) {}
-    ~CommonTypeSat() = default;
-
-    virtual void add(Satellite sat) { add_get_ucomp(sat); };
-    virtual void remove(Satellite sat) override;
-
-    std::tuple<UCOMP_T& ...> add_get_ucomp_all(Satellite sat);
-    auto& add_get_ucomp(Satellite sat);
-
-    constexpr Universe& get_universe() { return m_universe; }
-
-private:
-
-    Universe& m_universe;
-};
-
-template<typename TYPESAT_T, typename ... UCOMP_T>
-std::tuple<UCOMP_T& ...> CommonTypeSat<TYPESAT_T, UCOMP_T ...>::add_get_ucomp_all(Satellite sat)
+template<typename UCOMP_T>
+constexpr decltype(auto) emplace_filter_tags(Universe& rUni, Satellite sat)
```

Comment:
Could you document the goal behind returning an empty tuple if the requested type is empty?

Where is this used that that represents a big saving?

Reply:
Additionally, you can safely narrow the scope of your if-constexpr, i believe, to only the return line, to slightly cut down on duplicated lines.

Compiler will properly handle the reference variable being unused.

Reply:
Compilation fails when getting a reference to an empty struct, like UCompActiveArea. This is used by `add_get_ucomp_all`, which is suppose to emplace multiple components through variadic templates, and return a tuple of references all of them. This function together with `std::tuple_cat` can filter out the empty structs. It took a bit of trial and error to figure this out, but I learned quite a bit more metaprogramming on the way.

Reply:
I guess I'm just a bit sceptical that with a template function like this that the existence of the tag struct is actually kept at the assembly level.

But it's good that you learned more metaprogramming :-)

Reply:
> Compilation fails when getting a reference to an empty struct, like UCompActiveArea. This is used by add_get_ucomp_all,

Wait. What?

That's not true at all.

```
struct empty {};
void foo()
{
    empty bob;
    empty const& alice = bob;
}
```
That'll work just fine.

Reply:
```
template<typename UCOMP_T>
constexpr std::tuple<UCOMP_T&> emplace_filter_tags(Universe& rUni, Satellite sat)
{
    // Return reference tuple for normal struct
    return { rUni.get_reg().emplace<UCOMP_T>(sat) };
}
```

That should work just fine, regardless of the size of the types.

Reply:
Oh, this is an optimization by EnTT. `emplace` will return void if the type is an empty struct.

https://github.com/skypjack/entt/wiki/Crash-Course:-entity-component-system#empty-type-optimization

Reply:
Ahhhhhhhhh, that makes sense.